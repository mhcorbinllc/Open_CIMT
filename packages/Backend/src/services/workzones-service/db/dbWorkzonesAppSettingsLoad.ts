import {DynaMongoDB} from "dyna-mongo-db";
import {
  IWorkzonesAppSettings,
  defaultWorkzoneAppSettings,
} from "../interfaces/IWorkzonesAppSettings";
import {getWorkzonesAppSettingsCollection} from "./getWorkzonesCollections";

export const dbWorkzonesAppSettingsLoad = async (dmdb: DynaMongoDB, companyId: string, deviceId = '*'): Promise<IWorkzonesAppSettings> => {
  const collection = await getWorkzonesAppSettingsCollection(dmdb, companyId);
  const savedSettings = await collection.findOne({deviceId});

  if (savedSettings) {
    delete (savedSettings as any)._id;
    return savedSettings;
  }

  return {
    ...defaultWorkzoneAppSettings,
    deviceId,
    createdAt: Date.now(),
  };
};
