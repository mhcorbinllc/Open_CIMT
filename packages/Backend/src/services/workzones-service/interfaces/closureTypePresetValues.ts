import {
  EWZClosureType,
  EWZLane,
  EWZShoulder,
  IWorkzone,
} from "./IWorkzone";

export const closureTypePresetValues: Record<EWZClosureType, Partial<IWorkzone>> = {
  [EWZClosureType.NONE]: {
    itisCodes: [],
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 0,
  },
  [EWZClosureType.LEFT_LANE_CLOSED_MERGE_RIGHT]: {
    itisCodes: "1025 769 12545 8195 771 7451 13579".split(' ').map(Number),
    closedLane: EWZLane.LEFT,
    closedShoulder: EWZShoulder.LEFT,
    speedLimitInMiles: 0,
  },
  [EWZClosureType.RIGHT_LANE_CLOSED_MERGE_LEFT]: {
    itisCodes: "1025 769 12545 8196 771 7451 13580".split(' ').map(Number),
    closedLane: EWZLane.RIGHT,
    closedShoulder: EWZShoulder.RIGHT,
    speedLimitInMiles: 0,
  },
  [EWZClosureType.LEFT_SHOULDER_CLOSED]: {
    itisCodes: "1025 769 8209 771".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.LEFT,
    speedLimitInMiles: 0,
  },
  [EWZClosureType.RIGHT_SHOULDER_CLOSED]: {
    itisCodes: "1025 769 8208 771".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.RIGHT,
    speedLimitInMiles: 0,
  },
  [EWZClosureType.REDUCED_SPEED_15_MPH]: {
    itisCodes: "1025 268 12559 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 15,
  },
  [EWZClosureType.REDUCED_SPEED_20_MPH]: {
    itisCodes: "1025 268 12564 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 20,
  },
  [EWZClosureType.REDUCED_SPEED_25_MPH]: {
    itisCodes: "1025 268 12569 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 25,
  },
  [EWZClosureType.REDUCED_SPEED_30_MPH]: {
    itisCodes: "1025 268 12574 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 30,
  },
  [EWZClosureType.REDUCED_SPEED_35_MPH]: {
    itisCodes: "1025 268 12579 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 35,
  },
  [EWZClosureType.REDUCED_SPEED_40_MPH]: {
    itisCodes: "1025 268 12584 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 40,
  },
  [EWZClosureType.REDUCED_SPEED_45_MPH]: {
    itisCodes: "1025 268 12589 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 45,
  },
  [EWZClosureType.REDUCED_SPEED_50_MPH]: {
    itisCodes: "1025 268 12594 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 50,
  },
  [EWZClosureType.REDUCED_SPEED_55_MPH]: {
    itisCodes: "1025 268 12599 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 55,
  },
  [EWZClosureType.REDUCED_SPEED_60_MPH]: {
    itisCodes: "1025 268 12604 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 60,
  },
  [EWZClosureType.REDUCED_SPEED_65_MPH]: {
    itisCodes: "1025 268 12609 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 65,
  },
  [EWZClosureType.REDUCED_SPEED_70_MPH]: {
    itisCodes: "1025 268 12614 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 70,
  },
  [EWZClosureType.LEFT_TURN_SPEED_15_MPH]: {
    itisCodes: "13594 268 12559 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 15,
  },
  [EWZClosureType.LEFT_TURN_SPEED_20_MPH]: {
    itisCodes: "13594 268 12564 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 20,
  },
  [EWZClosureType.LEFT_TURN_SPEED_25_MPH]: {
    itisCodes: "13594 268 12569 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 25,
  },
  [EWZClosureType.LEFT_TURN_SPEED_30_MPH]: {
    itisCodes: "13594 268 12574 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 30,
  },
  [EWZClosureType.LEFT_TURN_SPEED_35_MPH]: {
    itisCodes: "13594 268 12579 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 35,
  },
  [EWZClosureType.LEFT_TURN_SPEED_40_MPH]: {
    itisCodes: "13594 268 12584 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 40,
  },
  [EWZClosureType.LEFT_TURN_SPEED_45_MPH]: {
    itisCodes: "13594 268 12589 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 45,
  },
  [EWZClosureType.LEFT_TURN_SPEED_50_MPH]: {
    itisCodes: "13594 268 12594 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 50,
  },
  [EWZClosureType.LEFT_TURN_SPEED_55_MPH]: {
    itisCodes: "13594 268 12599 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 55,
  },
  [EWZClosureType.RIGHT_TURN_SPEED_15_MPH]: {
    itisCodes: "13599 268 12559 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 15,
  },
  [EWZClosureType.RIGHT_TURN_SPEED_20_MPH]: {
    itisCodes: "13599 268 12564 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 20,
  },
  [EWZClosureType.RIGHT_TURN_SPEED_25_MPH]: {
    itisCodes: "13599 268 12569 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 25,
  },
  [EWZClosureType.RIGHT_TURN_SPEED_30_MPH]: {
    itisCodes: "13599 268 12574 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 30,
  },
  [EWZClosureType.RIGHT_TURN_SPEED_35_MPH]: {
    itisCodes: "13599 268 12579 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 35,
  },
  [EWZClosureType.RIGHT_TURN_SPEED_40_MPH]: {
    itisCodes: "13599 268 12584 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 40,
  },
  [EWZClosureType.RIGHT_TURN_SPEED_45_MPH]: {
    itisCodes: "13599 268 12589 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 45,
  },
  [EWZClosureType.RIGHT_TURN_SPEED_50_MPH]: {
    itisCodes: "13599 268 12594 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 50,
  },
  [EWZClosureType.RIGHT_TURN_SPEED_55_MPH]: {
    itisCodes: "13599 268 12599 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 55,
  },
  [EWZClosureType.HEIGHT_LIMIT_14_FT_0_IN]: {
    itisCodes: "2574 13569 12712 8709".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 0,
  },
};
