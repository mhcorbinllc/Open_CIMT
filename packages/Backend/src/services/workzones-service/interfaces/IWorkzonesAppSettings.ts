import {IWorkzone} from "./IWorkzone";
import {getDefaultAppItisCodes} from "./getDefaultAppItisCodes";
import {closureTypePresetValues} from "./closureTypePresetValues";
// App settings per company (not per device or user).

export interface IWorkzonesAppSettings {
  // See the apiWorkzonesAppSettingPost.ts's validation for the valid options per property

  deviceId: string;           // PK, "*" is for all devices of the company

  serviceChannel: string;
  psid: string;               // Courier font type
  txmode: EWZTXMode;
  intervalInMs: number;
  priority: number;           // 0..63
  useSignature: boolean;
  useEncryption: boolean;
  closureSelectionMode: EWZClosureSelectionMode;

  itisCodes: IItisCode[];

  createdAt: number;
  updatedAt: number;
  deletedAt: number;
}

export enum EWZTXMode {
  CONT = "CONT",
  ALT = "ALT",
}

export enum EWZClosureSelectionMode {
  CUSTOM = "CUSTOM",
  SIMPLE = "SIMPLE",
}

export interface IItisCode {
  code: number;
  label: string;
}

export const defaultWorkzoneAppSettings: IWorkzonesAppSettings = {
  deviceId: '',
  serviceChannel: "172",
  psid: '8002',
  txmode: EWZTXMode.CONT,
  intervalInMs: 1000,
  priority: 4,
  useSignature: false,
  useEncryption: false,
  closureSelectionMode: EWZClosureSelectionMode.SIMPLE,
  itisCodes: getDefaultAppItisCodes(),
  createdAt: 0,
  updatedAt: 0,
  deletedAt: 0,
};

export interface IValidationErrorItisRemoveBlockedByWorkzone {
  removedItisCodes: number[];
  workzone: IWorkzone;
}

export interface IClosureTypeRemovedItisCodes {
  closureType: string;
  removedItisCodes: number[];
}

export const getRemovedSimpleClosureTypeItisCodes  = (
  itisCodes: IItisCode[],
): IClosureTypeRemovedItisCodes[] => {
  const closureTypeRemovedItisCodes: IClosureTypeRemovedItisCodes[] = [];
  const codesList = itisCodes.map(itisCode => (itisCode.code));
  Object.keys(closureTypePresetValues)
    .forEach((key) => {
      const newRecord: IClosureTypeRemovedItisCodes = {
        closureType: key,
        removedItisCodes: [],
      };
      closureTypePresetValues[key].itisCodes
        .forEach((cti: number) => {
          const exists = codesList.includes(cti);
          if (exists) return;
          newRecord.removedItisCodes.push(cti);
        });
      closureTypeRemovedItisCodes.push(newRecord);
    });

  return closureTypeRemovedItisCodes.filter(closureType => closureType.removedItisCodes.length > 0);
};
