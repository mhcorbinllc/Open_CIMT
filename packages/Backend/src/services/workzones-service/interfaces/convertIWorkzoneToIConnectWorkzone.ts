import {IWorkzone} from "./IWorkzone";
import {IConnectWorkzone} from "./IConnectWorkzone";
import {IWorkzonesAppSettings} from "./IWorkzonesAppSettings";

export const convertIWorkzoneToIConnectWorkzone = (
  {
    workzone,
    workzoneAppSettings,
  }: {
    workzone: IWorkzone;
    workzoneAppSettings: IWorkzonesAppSettings;
  },
): IConnectWorkzone => {
  return {
    id: workzone.kapschId,
    name: workzone.name,
    speedlimit: Math.round(workzone.speedLimitInMiles),
    isWorkzone: true,
    areWorkersPresent: workzone.workersPresent,
    shoulder: workzone.closedShoulder,
    lane: workzone.closedLane,
    laneWidth: Math.round(workzone.closedLaneWidthInMeters),
    start: dateToIsoString(workzone.start),
    end: dateToIsoString(workzone.end),
    forEver: workzone.forEver,
    itisCodes: workzone.itisCodes,

    // Point
    point: {
      type: "Point",
      coordinates: [
        workzone.point.lng,
        workzone.point.lat,
        workzone.point.alt === undefined
          ? null
          : workzone.point.alt,
      ],
    },

    heading: workzone.heading,

    // Path
    usePath: !!workzone.path.points.length,
    path: {
      type: "LineString",
      coordinates:
        workzone.path.points
          .map(point => ([
            point.lng,
            point.lat,
            point.alt || 0,
          ])),
    },

    // Selection mode: Region
    broadcastRegion: {
      type: "Polygon",
      coordinates: [
        workzone.broadcast.selectionByRegion.region.points
          .map(point => ([
            point.lng,
            point.lat,
            point.alt || 0,
          ])),
      ],
    },

    // Selection mode: Radius
    useRadius: !!workzone.broadcast.selectionByRadius.radiusInMeters,
    radius: workzone.broadcast.selectionByRadius.radiusInMeters,

    // CIMs features options
    serviceChannelStr: workzoneAppSettings.serviceChannel,
    psid: workzoneAppSettings.psid,
    txmode: workzoneAppSettings.txmode,
    interval: workzoneAppSettings.intervalInMs,
    priority: workzoneAppSettings.priority,
    useSignature: workzoneAppSettings.useSignature,
    useEncryption: workzoneAppSettings.useEncryption,
  };
};

const dateToIsoString = (date: Date | string | number): string => {
  return new Date(date).toISOString()
    .split('.')[0] + "Z";
};
