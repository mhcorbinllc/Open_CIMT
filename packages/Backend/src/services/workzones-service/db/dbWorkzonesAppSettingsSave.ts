import {DynaMongoDB} from "dyna-mongo-db";
import {IWorkzonesAppSettings} from "../interfaces/IWorkzonesAppSettings";
import {getWorkzonesAppSettingsCollection} from "./getWorkzonesCollections";

export const dbWorkzonesAppSettingsSave = async (
  {
    dmdb,
    companyId,
    workzoneAppSettings,
    externalImport = false,
    newerOnly = false,
  }: {
    dmdb: DynaMongoDB;
    companyId: string;
    workzoneAppSettings: IWorkzonesAppSettings;
    externalImport?: boolean;
    newerOnly?: boolean;
  },
): Promise<void> => {
  const collection = await getWorkzonesAppSettingsCollection(dmdb, companyId);

  const currentWorkzoneAppSettings = await collection.findOne({deviceId: workzoneAppSettings.deviceId});

  if (
    newerOnly
    && currentWorkzoneAppSettings
    && currentWorkzoneAppSettings.updatedAt > workzoneAppSettings.updatedAt
  ) {
    return;
  }

  const now = Date.now();

  await collection.updateOne(
    {deviceId: workzoneAppSettings.deviceId},
    {
      $set:
        externalImport
          ? workzoneAppSettings
          : {
            ...workzoneAppSettings,
            createdAt: currentWorkzoneAppSettings ? currentWorkzoneAppSettings.createdAt : now,
            updatedAt: now,
            deletedAt: 0,
          },
    },
    {upsert: true},
  );
};
