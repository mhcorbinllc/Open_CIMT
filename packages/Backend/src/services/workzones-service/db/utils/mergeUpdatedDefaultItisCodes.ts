import {Db} from "mongodb";
import {syncPromises} from "dyna-sync";

import {
  IWorkzonesAppSettings,
} from "../../interfaces/IWorkzonesAppSettings";
import {getDefaultAppItisCodes} from "../../interfaces/getDefaultAppItisCodes";

export const mergeUpdatedDefaultItisCodes = async (
  {
    db,
    collectionName,
  }: {
    db: Db;
    collectionName: string;
  },
): Promise<void> => {
  const collection = db.collection<IWorkzonesAppSettings>(collectionName);
  const appsSettings = await collection.find().toArray();
  const defaultAppItisCodes = getDefaultAppItisCodes();

  await syncPromises(
    ...appsSettings.map(appSettings => {

      const hasSameCodes =
        appSettings.itisCodes.map(v => v.code).sort()
          .join(',') ===
        defaultAppItisCodes.map(v => v.code).sort()
          .join(',');

      if (hasSameCodes) return () => Promise.resolve();

      const mergedItisCodes = defaultAppItisCodes
        .filter((v) => !appSettings.itisCodes.find(t =>
          t.code === v.code
          && t.label.toLowerCase() === v.label.toLowerCase(),
        ))
        .concat(appSettings.itisCodes);

      return async () => {
        await collection.updateOne(
          {deviceId: appSettings.deviceId},
          {$set: {itisCodes: mergedItisCodes}},
        );
      };
    }),
  );
};
