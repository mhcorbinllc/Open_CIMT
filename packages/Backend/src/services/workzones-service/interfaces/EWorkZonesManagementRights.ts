import {convertDictionaryToViewLabelArray} from "../../../utils/convertDictionaryToViewLabelArray";

export enum EWorkZonesManagementRights {
  WORKZONES_VIEW = "WZM_V",
  WORKZONES_EDIT = "WZM_E",
}

const labelsDic: Record<EWorkZonesManagementRights, string>  = {
  [EWorkZonesManagementRights.WORKZONES_VIEW]: 'CIMs / View',
  [EWorkZonesManagementRights.WORKZONES_EDIT]: 'CIMs / Edit',
};

export const EWorkZonesManagementRightsArray = convertDictionaryToViewLabelArray(labelsDic);
