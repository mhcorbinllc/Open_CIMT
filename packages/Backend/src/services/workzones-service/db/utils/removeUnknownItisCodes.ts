import {IDBWorkzone} from "../../interfaces/IDBWorkzone";
import {DynaMongoDB} from "dyna-mongo-db";
import {dbWorkzonesAppSettingsLoad} from "../dbWorkzonesAppSettingsLoad";

export const removeUnknownItisCodes = async (
  {
    dmdb,
    companyId,
    dbWorkzones,
  }: {
    dmdb: DynaMongoDB;
    companyId: string;
    dbWorkzones: IDBWorkzone[];
  },
): Promise<void> => {
  const settings = await dbWorkzonesAppSettingsLoad(dmdb, companyId);
  dbWorkzones
    .forEach(dbWorkzone => {
      dbWorkzone.itisCodes =
        dbWorkzone.itisCodes
          .filter(itisCode => settings.itisCodes.find(ic => itisCode === ic.code));
    });
};
