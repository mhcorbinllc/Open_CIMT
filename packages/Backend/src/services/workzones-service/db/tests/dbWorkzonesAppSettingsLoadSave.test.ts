import {testDMDB} from "../../../../../tests/utils/testDMDB";
import {dbWorkzonesAppSettingsSave} from "../dbWorkzonesAppSettingsSave";
import {EWZTXMode} from "../../interfaces/IWorkzonesAppSettings";
import {dbWorkzonesAppSettingsLoad} from "../dbWorkzonesAppSettingsLoad";
import {clearDynamicValuesForSnapshots} from "../../../../utils/clearDynamicValuesForSnapshots";

describe('CIMT - App Settings', () => {
  const companyId = 'test-company--test-85739524';

  const c = (data: any): any => clearDynamicValuesForSnapshots(data);

  const dmdb = testDMDB;
  const clearDb = async () => {
    try {
      await dmdb.dropCollection([companyId, 'workzones-app-settings'].join('---'));
    }
    catch (e) {
      console.error('clearDb error:', e);
    }
  };

  beforeAll(async () => {
    await clearDb();
  });
  afterAll(async () => {
    await clearDb();
    await dmdb.disconnect();
  });

  test('Load default', async () => {
    const settings = await dbWorkzonesAppSettingsLoad(dmdb, companyId);
    expect(c(settings)).toMatchSnapshot();
  });

  test('Load default, edit, save load 1', async () => {
    const workzoneAppSettings = await dbWorkzonesAppSettingsLoad(dmdb, companyId);
    workzoneAppSettings.itisCodes.push({
      code: 13,
      label: 'Fab',
    });
    await dbWorkzonesAppSettingsSave({
      dmdb,
      companyId,
      workzoneAppSettings,
    });
    const loadedSettings = await dbWorkzonesAppSettingsLoad(dmdb, companyId);
    expect(c(loadedSettings)).toMatchSnapshot();
  });

  test('Load default, edit, save load 2', async () => {
    const settings = await dbWorkzonesAppSettingsLoad(dmdb, companyId);
    await dbWorkzonesAppSettingsSave({
      dmdb,
      companyId,
      workzoneAppSettings: {
        ...settings,
        serviceChannel: "200",
        psid: 'PSID-2423',
        txmode: EWZTXMode.CONT,
        intervalInMs: 1200,
        priority: 12,
        useSignature: true,
        useEncryption: true,
        itisCodes: [
          ...settings.itisCodes,
          {
            code: 14,
            label: 'Conform',
          },
        ],
      },
    });
    const loadedSettings = await dbWorkzonesAppSettingsLoad(dmdb, companyId);
    expect(c(loadedSettings)).toMatchSnapshot();
  });
});
