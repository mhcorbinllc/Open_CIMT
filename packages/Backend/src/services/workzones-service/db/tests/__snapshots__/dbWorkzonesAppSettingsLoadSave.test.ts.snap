// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CIMT - App Settings Load default 1`] = `
Object {
  "closureSelectionMode": "SIMPLE",
  "createdAt": "---NUMBER>0---",
  "deletedAt": "---NUMBER=0---",
  "deviceId": "*",
  "intervalInMs": 1000,
  "itisCodes": Array [
    Object {
      "code": 1025,
      "label": "Road Construction",
    },
    Object {
      "code": 769,
      "label": "Closed to Traffic",
    },
    Object {
      "code": 12545,
      "label": "1",
    },
    Object {
      "code": 8196,
      "label": "Right Lane",
    },
    Object {
      "code": 8195,
      "label": "Left Lane",
    },
    Object {
      "code": 771,
      "label": "Closed Ahead",
    },
    Object {
      "code": 7451,
      "label": "Merge",
    },
    Object {
      "code": 13580,
      "label": "Left",
    },
    Object {
      "code": 13579,
      "label": "Right",
    },
    Object {
      "code": 8208,
      "label": "Right Shoulder",
    },
    Object {
      "code": 8209,
      "label": "Left Shoulder",
    },
    Object {
      "code": 13594,
      "label": "Left Turn",
    },
    Object {
      "code": 13599,
      "label": "Right Turn",
    },
    Object {
      "code": 268,
      "label": "Speed Limit",
    },
    Object {
      "code": 12559,
      "label": "15",
    },
    Object {
      "code": 12564,
      "label": "20",
    },
    Object {
      "code": 12569,
      "label": "25",
    },
    Object {
      "code": 12574,
      "label": "30",
    },
    Object {
      "code": 12579,
      "label": "35",
    },
    Object {
      "code": 12584,
      "label": "40",
    },
    Object {
      "code": 12589,
      "label": "45",
    },
    Object {
      "code": 12594,
      "label": "50",
    },
    Object {
      "code": 12599,
      "label": "55",
    },
    Object {
      "code": 12604,
      "label": "60",
    },
    Object {
      "code": 12609,
      "label": "65",
    },
    Object {
      "code": 12614,
      "label": "70",
    },
    Object {
      "code": 8720,
      "label": "MPH",
    },
    Object {
      "code": 2574,
      "label": "Height Limit",
    },
    Object {
      "code": 13569,
      "label": "Ahead",
    },
    Object {
      "code": 12712,
      "label": "168",
    },
    Object {
      "code": 8709,
      "label": "Inches",
    },
  ],
  "priority": 4,
  "psid": "8002",
  "serviceChannel": "172",
  "txmode": "CONT",
  "updatedAt": "---NUMBER=0---",
  "useEncryption": false,
  "useSignature": false,
}
`;

exports[`CIMT - App Settings Load default, edit, save load 1 1`] = `
Object {
  "closureSelectionMode": "SIMPLE",
  "createdAt": "---NUMBER>0---",
  "deletedAt": "---NUMBER=0---",
  "deviceId": "*",
  "intervalInMs": 1000,
  "itisCodes": Array [
    Object {
      "code": 1025,
      "label": "Road Construction",
    },
    Object {
      "code": 769,
      "label": "Closed to Traffic",
    },
    Object {
      "code": 12545,
      "label": "1",
    },
    Object {
      "code": 8196,
      "label": "Right Lane",
    },
    Object {
      "code": 8195,
      "label": "Left Lane",
    },
    Object {
      "code": 771,
      "label": "Closed Ahead",
    },
    Object {
      "code": 7451,
      "label": "Merge",
    },
    Object {
      "code": 13580,
      "label": "Left",
    },
    Object {
      "code": 13579,
      "label": "Right",
    },
    Object {
      "code": 8208,
      "label": "Right Shoulder",
    },
    Object {
      "code": 8209,
      "label": "Left Shoulder",
    },
    Object {
      "code": 13594,
      "label": "Left Turn",
    },
    Object {
      "code": 13599,
      "label": "Right Turn",
    },
    Object {
      "code": 268,
      "label": "Speed Limit",
    },
    Object {
      "code": 12559,
      "label": "15",
    },
    Object {
      "code": 12564,
      "label": "20",
    },
    Object {
      "code": 12569,
      "label": "25",
    },
    Object {
      "code": 12574,
      "label": "30",
    },
    Object {
      "code": 12579,
      "label": "35",
    },
    Object {
      "code": 12584,
      "label": "40",
    },
    Object {
      "code": 12589,
      "label": "45",
    },
    Object {
      "code": 12594,
      "label": "50",
    },
    Object {
      "code": 12599,
      "label": "55",
    },
    Object {
      "code": 12604,
      "label": "60",
    },
    Object {
      "code": 12609,
      "label": "65",
    },
    Object {
      "code": 12614,
      "label": "70",
    },
    Object {
      "code": 8720,
      "label": "MPH",
    },
    Object {
      "code": 2574,
      "label": "Height Limit",
    },
    Object {
      "code": 13569,
      "label": "Ahead",
    },
    Object {
      "code": 12712,
      "label": "168",
    },
    Object {
      "code": 8709,
      "label": "Inches",
    },
    Object {
      "code": 13,
      "label": "Fab",
    },
  ],
  "priority": 4,
  "psid": "8002",
  "serviceChannel": "172",
  "txmode": "CONT",
  "updatedAt": "---NUMBER>0---",
  "useEncryption": false,
  "useSignature": false,
}
`;

exports[`CIMT - App Settings Load default, edit, save load 2 1`] = `
Object {
  "closureSelectionMode": "SIMPLE",
  "createdAt": "---NUMBER>0---",
  "deletedAt": "---NUMBER=0---",
  "deviceId": "*",
  "intervalInMs": 1200,
  "itisCodes": Array [
    Object {
      "code": 1025,
      "label": "Road Construction",
    },
    Object {
      "code": 769,
      "label": "Closed to Traffic",
    },
    Object {
      "code": 12545,
      "label": "1",
    },
    Object {
      "code": 8196,
      "label": "Right Lane",
    },
    Object {
      "code": 8195,
      "label": "Left Lane",
    },
    Object {
      "code": 771,
      "label": "Closed Ahead",
    },
    Object {
      "code": 7451,
      "label": "Merge",
    },
    Object {
      "code": 13580,
      "label": "Left",
    },
    Object {
      "code": 13579,
      "label": "Right",
    },
    Object {
      "code": 8208,
      "label": "Right Shoulder",
    },
    Object {
      "code": 8209,
      "label": "Left Shoulder",
    },
    Object {
      "code": 13594,
      "label": "Left Turn",
    },
    Object {
      "code": 13599,
      "label": "Right Turn",
    },
    Object {
      "code": 268,
      "label": "Speed Limit",
    },
    Object {
      "code": 12559,
      "label": "15",
    },
    Object {
      "code": 12564,
      "label": "20",
    },
    Object {
      "code": 12569,
      "label": "25",
    },
    Object {
      "code": 12574,
      "label": "30",
    },
    Object {
      "code": 12579,
      "label": "35",
    },
    Object {
      "code": 12584,
      "label": "40",
    },
    Object {
      "code": 12589,
      "label": "45",
    },
    Object {
      "code": 12594,
      "label": "50",
    },
    Object {
      "code": 12599,
      "label": "55",
    },
    Object {
      "code": 12604,
      "label": "60",
    },
    Object {
      "code": 12609,
      "label": "65",
    },
    Object {
      "code": 12614,
      "label": "70",
    },
    Object {
      "code": 8720,
      "label": "MPH",
    },
    Object {
      "code": 2574,
      "label": "Height Limit",
    },
    Object {
      "code": 13569,
      "label": "Ahead",
    },
    Object {
      "code": 12712,
      "label": "168",
    },
    Object {
      "code": 8709,
      "label": "Inches",
    },
    Object {
      "code": 13,
      "label": "Fab",
    },
    Object {
      "code": 14,
      "label": "Conform",
    },
  ],
  "priority": 12,
  "psid": "PSID-2423",
  "serviceChannel": "200",
  "txmode": "CONT",
  "updatedAt": "---NUMBER>0---",
  "useEncryption": true,
  "useSignature": true,
}
`;
